kernel/kernel-api/src/main/java/io/delta/kernel/internal/util/SchemaUtils.java

```diff
@@ -436,6 +440,123 @@ private static Map<Long, StructField> fieldsById(StructType schema) {
     return columnIdToField;
   }
 
+  private static void validateNestedIds(
+      SchemaChanges schemaChanges,
+      Map<Long, StructField> currentFieldsById,
+      Map<Long, StructField> updatedFieldsById) {
+    for (Tuple2<StructField, StructField> updatedField : schemaChanges.updatedFields()) {
+      validateNestedIdsExist(updatedField._2);
+    }
+
+    for (StructField addedField : schemaChanges.addedFields()) {
+      validateNestedIdsExist(addedField);
+    }
+
+    validateNestedIdConsistency(currentFieldsById, updatedFieldsById);
+  }
+
+  private static void validateNestedIdsExist(StructField field) {
+    if (field.getDataType() instanceof MapType || field.getDataType() instanceof ArrayType) {
+      checkArgument(
+          hasNestedColumnIds(field),
+          "Field %s is missing expected nested column IDs",
+          field.getName());
+      if (field.getDataType() instanceof MapType) {
+        checkArgument(
+            getNestedFieldIds(field).size() >= 2,
+            "Map field %s needs at least 2 nested IDs, one for the key and one for the value",
+            field.getName());
+      }
+    }
+  }
+
+  private static Map<Long, List<NestedIdEntry>> columnIdToNestedIds(
+      Map<Long, StructField> currentFieldIdToField) {
+    Map<Long, List<NestedIdEntry>> nestedIdEntries = new HashMap<>();
+    for (Map.Entry<Long, StructField> entry : currentFieldIdToField.entrySet()) {
+      if (hasNestedColumnIds(entry.getValue())) {
+        nestedIdEntries.put(entry.getKey(), getNestedFieldIds(entry.getValue()));
+      }
+    }
+
+    return nestedIdEntries;
+  }
+
+  /**
+   * Validate that any nested IDs that were in the current field are unchanged with the ones in the
+   * updated field and that the nested IDs are not duplicate
+   */
+  private static void validateNestedIdConsistency(
+      Map<Long, StructField> currentFieldsById, Map<Long, StructField> updatedFieldsById) {
+    Map<Long, List<NestedIdEntry>> currentFieldIdToNestedIds =
+        columnIdToNestedIds(currentFieldsById);
+    Map<Long, List<NestedIdEntry>> updatedFieldIdToNestedIds =
+        columnIdToNestedIds(updatedFieldsById);
+    /* Validate that nested field IDs exist in the schema
+    and are consistent with the previous schemas */
+    for (Map.Entry<Long, List<NestedIdEntry>> entry : updatedFieldIdToNestedIds.entrySet()) {
```

Comment:
Wondering if we can/should use the `SchemaChanges` here to check updated fields since it seems applicable and a useful breakdown?

Reply:
Yeah I originally punted on using the SchemaChanges here because we needed the grouping by field ID for the nested ID consistency check and SchemaChanges currently is agnostic to that (doesn't care about how the changes are produced, that's up to the person building the changes). 

But I just realized we probably could build that grouping ourselves from the changes, and still just go through the updates and added columns ourselves for now and filter out the entries we want to do the nested ID check. That'll save the extra unnecessary validations against unchanged fields.

Reply:
what do you mean by "grouping by field ID"?

Isn't the schemChanges updates determined via fieldId -- so we have oldField and newField (same fieldId) and can just verify the metadata?

Reply:
That's what's happening now. I just meant that since SchemaChanges itself wasn't aware of grouping by field ID, we have to extract the ID from the changes again to be able to figure out which fields we want do the nested ID validation on